draft-irtf-cfrg-hpke.md

```diff
@@ -300,23 +304,33 @@ operations, roles, and behaviors of HPKE:
 HPKE variants rely on the following primitives:
 
 * A Key Encapsulation Mechanism (KEM):
-  - GenerateKeyPair(): Generate a key pair (skX, pkX)
-  - DeriveKeyPair(ikm): Derive a key pair `(skX, pkX)` from the byte string `ikm`,
-    where `ikm` SHOULD have at least `Nsk` bytes of entropy (see
-    {{derive-key-pair}} for discussion)
-  - Serialize(pk): Produce a byte string of length `Npk` encoding the
-    public key `pk`
-  - Deserialize(enc): Parse the byte string `enc` of length `Npk` to recover a
-    public key (note: this function can raise an error upon `enc` deserialization failure)
-  - Encap(pk): Generate an ephemeral, fixed-length symmetric key (the KEM shared secret) and
-    a fixed-length encapsulation of that key that can be decapsulated
-    by the holder of the private key corresponding to pk
+  - GenerateKeyPair() (OPTIONAL): Generate a key pair `(skX, pkX)`
```

Comment:
Hmm, this seems mandatory since it's used in Encap, no? Is the idea that future Encaps might not have a key generation function? If not, how would keys, well, be generated?

Reply:
For DHKEM it is mandatory because it is used in Encap/Decap as you say. Other KEMs, and RSA for example, do not use it in Encap. For these KEMs, the HPKE implementation can receive the keys from outside (the higher-level application). The keys would then also be assumed to be generated by the application. The RFCs mentioning RSA KEMs https://tools.ietf.org/html/rfc5990 and https://tools.ietf.org/html/rfc8017 for example do not mention key generation at all.

Reply:
Having said that, I agree that saying `OPTIONAL` if the DHKEMs defined in the document definitely require it, is odd. We could write optional in lowercase, or don't mention it in the Cryptographic Dependencies section at all and only leave the explanation in the Future KEMs section. Do you have a preference?

Reply:
I reverted all my changes related to making these functions optional, because:
- For GenerateKeyPair, any KEM (and RSA too) can just write something like “GenerateKeyPair is defined as in [reference]” and fullfil the requirement easily like this.
- DeriveKeyPair(ikm): here, I see two possibilities:
  * any KEM (and also RSA) could just pass a complete keypair `sk, pk` to the function and let the function be the identity function
  * any KEM (and also RSA) could pass whatever encoding of a complete keypair to the function and let the function be the deserialization of this encoding.

Do we need to mention this in the document?

Reply:
I don't think it's necessary. Thanks!